### ISSUE TITLE
Switch to kingpin?

### ISSUE THREAD

**Issue by jovandeginste** [drakkan/sftpgo#20]
For argument parsing in my golang projects, I use [kingpin](https://godoc.org/github.com/alecthomas/kingpin). Would you be interested in a PR for switching to this library?

**Comment by drakkan** [comment]
argument parsing is quite simple now (no subcommand ecc..) and we should use arguments only for configurations that cannot be stored inside `sftpgo.conf` for example config dir and logging parameters (we need to inizialize the logger to be able to log a configuration error).

I hope we'll not add other arguments, anyway if a more advanced library such as kingpin or cobra can give significative adavantages over using the standard  `flag` a contribution is welcome, thanks! 

Do you think to add arguments for other things? For example a useful addition would be a cli to use the REST api. If we add this directly inside sftpgo we need a more advanced argument parsing library, but I was thinking about adding a python script in a `scripts` or `samples` directory to do this. I would like to keep the main sftpgo application as clean as possible

**Comment by jovandeginste** [comment]
I agree it's simple and should remain simple, but some advantages of eg. kingpin is automatic generation of help/man pages, shell completion (not the most relevant for a daemon) and eg. coupling between flags, env vars and config file (override a config file setting through an argument flag).

In the world of containers, config files are cumbersome and arguments are easier...

About the REST api, I would write a dedicated cli tool, and I would write it in Go :-)

**Comment by drakkan** [comment]
override a config file setting through an argument flag or an env var it's interesting, if this does not complicate the code much then a pull request is welcome, thanks!

**Comment by drakkan** [comment]
default values for arguments can now be set from env vars:

https://github.com/drakkan/sftpgo/commit/ba3f9d891a93f68edfcf58a4369cb869b4d8a6b2

I think we need a dedicated arguments parser only if we need subcommands and/or other advanced features

**Comment by drakkan** [comment]
I did a quick test using cobra, here is the result:

```
./sftpgo 
Full featured and highly configurable SFTP server

Usage:
  sftpgo [command]

Available Commands:
  help        Help about any command
  serve       Start the SFTP Server

Flags:
  -h, --help   help for sftpgo

Use "sftpgo [command] --help" for more information about a command.
```

```
./sftpgo serve --help
To start the SFTP Server with the defaults values for the command line flags simply use:

sftpgo serve
		
Please take a look at the usage below to customize the startup options

Usage:
  sftpgo serve [flags]

Flags:
  -c, --config-dir string         Location for SFTPGo config dir. It must contain sftpgo.conf or the configured config-file-name and it is used as the base for files with a relative path (eg. the private keys for the SFTP server, the SQLite database if you use SQLite as data provider). (default ".")
  -f, --config-file-name string   Name for SFTPGo configuration file. It must be the name of a file stored in config-dir not the absolute path to the configuration file (default "sftpgo.conf")
  -h, --help                      help for serve
  -z, --log-compress              Determine if the rotated log files should be compressed using gzip
  -l, --log-file-path string      Location for the log file (default "sftpgo.log")
  -a, --log-max-age int           Maximum number of days to retain old log files (default 28)
  -b, --log-max-backups int       Maximum number of old log files to retain (default 5)
  -s, --log-max-size int          Maximum size in megabytes of the log file before it gets rotated. (default 10)
  -v, --log-verbose               Enable verbose logs (default true)
```

honestly I don't know which one to prefer, cobra seems to have more users and seems used in big projects. It has many depencencies but the final binary size is smaller than the one obtained using kingpin. Regarging kingpin I don't like that the defaults are always string, but this is minor and probably a personal taste. 

Cobra is also well integrated with viper  (#24) if we need to use a more advanced configuration library (I think it is not needed for now)

Did you use cobra or any other arguments parser library too? Why do you prefer kingpin? Thanks

**Comment by jovandeginste** [comment]
Regarding defaults being string, this is only the "input" (because on the CLI, everything is a string). It is then parsed to whatever format you want (and is supported by kingpin).

So:

```golang
logVerbose    = kingpin.Flag("log-verbose", "Enable verbose logs").Default("true").Bool()
```

... has an actual `bool` as result.

I'm not sure anymore why I picked kingpin when I first needed such a library, but I'm pretty sure I tried cobra back then and there must have been some issue for me that was easily solved with kingpin... I believe kingpin is more idiomatic and less intrusive in your code? Just a feeling :-)

OTOH, I have no horse in the game, so pick whatever you feel is best, and I'll be happy to contribute what I can when we have other unmet needs.
